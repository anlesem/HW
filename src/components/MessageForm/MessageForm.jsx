import { useRef, useEffect } from 'react';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import PropTypes from 'prop-types';

import { getTempInput } from '../../store/messages/selectors';
import { changeTempInput } from '../../store/messages/actions';

import style from './MessageForm.module.scss';

import Button from '@mui/material/Button';
import SendIcon from '@mui/icons-material/Send';
import TextField from '@mui/material/TextField';

export const MessageForm = ({ data }) => {
  const dispatch = useDispatch();
  const tempInput = useSelector(getTempInput, shallowEqual);
  const focusForm = useRef(null);
  const id = +data.chatId;
  const input = tempInput[id];

  const handleSubmit = (event) => {
    event.preventDefault();
    if (input) {
      data.sendMassage(input);
      dispatch(changeTempInput(id, ''));
    }
  };

  const handleInput = (value) => {
    dispatch(changeTempInput(id, value));
  };

  const handleKeyDown = (event) => {
    if (event.keyCode === 13 && !event.shiftKey) {
      event.preventDefault();
      handleSubmit(event);
    }
  };

  useEffect(() => {
    focusForm.current.focus();
  });

  return (
    <form className={style.form} onSubmit={handleSubmit}>
      <TextField
        id="outlined-multiline-static"
        name="text"
        label="Текст сообщения"
        multiline
        rows={2}
        value={input}
        onChange={(event) => handleInput(event.target.value)}
        onKeyDown={(event) => handleKeyDown(event)}
        inputRef={focusForm}
        disabled={data.chatId > 0 ? false : true}
        required
        inputProps={{ 'data-testid': 'message-form-input' }}
        className={style.textField}
      />
      <Button
        variant="outlined"
        type="submit"
        disabled={data.chatId > 0 ? false : true}
        data-testid="message-form-button">
        <SendIcon />
      </Button>
    </form>
  );
};

MessageForm.propTypes = {
  data: PropTypes.shape({
    chatId: PropTypes.string.isRequired,
    sendMassage: PropTypes.func.isRequired
  })
};
