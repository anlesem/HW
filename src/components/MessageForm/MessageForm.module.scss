.form {
   width: 100%;
   background-color: #fff;

   display: flex;
   justify-content: space-between;

   button {
      margin-left: 10px;
   }
}

.textField {
   flex-grow: 1;
}