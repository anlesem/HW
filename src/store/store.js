import { createStore, combineReducers, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import { profileReducer } from './profile/reducer';
import { chatsReducer } from './chats/reducer';
import { messagesReducer } from './messages/reducer';
import { dataReducer } from './data/reducer';

const persistConfig = {
  key: 'root',
  storage,
  blacklist: ['data']
};

const rootReducers = combineReducers({
  profile: profileReducer,
  chats: chatsReducer,
  messages: messagesReducer,
  data: dataReducer
});

const persistedReducers = persistReducer(persistConfig, rootReducers);

export const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const store = createStore(persistedReducers, composeEnhancers(applyMiddleware(thunk)));

export const persistor = persistStore(store);
