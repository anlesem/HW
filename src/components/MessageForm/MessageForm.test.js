import { render, screen, fireEvent } from '@testing-library/react';
import { MessageForm } from './MessageForm';

import { Provider } from 'react-redux';
import { store } from '../../store/store';

describe('MessageForm', () => {
  it('Компонент существует', () => {
    expect(MessageForm).toBeInstanceOf(Function);
  });

  it('Снимок состояния', () => {
    const chatId = '1';
    const sendMassage = jest.fn();
    const { asFragment } = render(
      <Provider store={store}>
        <MessageForm data={{ chatId, sendMassage }} />
      </Provider>
    );
    expect(asFragment(<MessageForm data={{ chatId, sendMassage }} />)).toMatchSnapshot();
  });

  it('Начало: Основное поле ввода пустое', () => {
    const chatId = '1';
    const sendMassage = jest.fn();
    render(
      <Provider store={store}>
        <MessageForm data={{ chatId, sendMassage }} />
      </Provider>
    );

    const input = screen.getByTestId('message-form-input');
    expect(input.value).toBe('');
  });

  it('Заполнение поле ввода', () => {
    const chatId = '1';
    const sendMassage = jest.fn();
    render(
      <Provider store={store}>
        <MessageForm data={{ chatId, sendMassage }} />
      </Provider>
    );

    const input = screen.getByTestId('message-form-input');

    fireEvent.change(input, { target: { value: 'bla-bla-bla' } });
    expect(input.value).toBe('bla-bla-bla');
  });

  it('Обработка клика на кнопку', () => {
    const chatId = '1';
    const sendMassage = jest.fn();

    render(
      <Provider store={store}>
        <MessageForm data={{ chatId, sendMassage }} />
      </Provider>
    );

    const input = screen.getByTestId('message-form-input');
    const button = screen.getByTestId('message-form-button');

    fireEvent.change(input, { target: { value: 'bla-bla-bla' } });
    expect(input.value).toBe('bla-bla-bla');

    fireEvent.click(button);
    expect(sendMassage).toBeCalled();
    expect(input.value).toBe('');
  });
});
