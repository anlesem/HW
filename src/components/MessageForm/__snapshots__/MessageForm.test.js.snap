// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageForm Снимок состояния 1`] = `
<DocumentFragment>
  <form
    class="form"
  >
    <div
      class="MuiFormControl-root MuiTextField-root textField css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-focused Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="outlined-multiline-static"
        id="outlined-multiline-static-label"
      >
        Текст сообщения
        <span
          aria-hidden="true"
          class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl MuiInputBase-multiline css-dpjnhs-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <textarea
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
          data-testid="message-form-input"
          id="outlined-multiline-static"
          name="text"
          required=""
          rows="2"
          style="height: 0px; overflow: hidden;"
        />
        <textarea
          aria-hidden="true"
          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
          readonly=""
          style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
          tabindex="-1"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1z7n62"
          >
            <span>
              Текст сообщения *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
      data-testid="message-form-button"
      tabindex="0"
      type="submit"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="SendIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M2.01 21 23 12 2.01 3 2 10l15 2-15 2z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
